### ISSUE TITLE
Speech translation on a custom dataset

### ISSUE THREAD

**Issue by balag59** [espnet/espnet#1904]
Hi,
I had a few queries regarding speech translation:
1. How would one go about training a ST model on a custom dataset with source audio and target text? What format does the data need to be in and what forms of pre-processing would be necessary?
2. Can I use one pre-trained ST model to fine-tune another ST model on a new language pair? Instead of starting ST training from scratch, I wanted to know if it's possible to fine-tune an existing ST model. I'm aware that I can use pretrained ASR and MT models for ST training, but I was wondering if I could directly fine-tune an existing model. 
3. Ideally, how many hours of speech data would I need to get a ST model working reasonably well?Will this amount reduce if I'm finetuning on an existing ST model instead of training from scratch?
4. Could you please guide me to the paper/architecture for the transformer based ST model used in Spanish-English ST task? I was unable to find this and it would be very helpful to know.

Thanks!

**Comment by sw005320** [comment]
@hirofumi0810, could you answer these questions?

**Comment by hirofumi0810** [comment]
@balag59
If you want to get better models, you need a triplet (source audio, source text, target text).
> 1. How would one go about training a ST model on a custom dataset with source audio and target text? What format does the data need to be in and what forms of pre-processing would be necessary?


I'm not sure ST models can be transferred to other language pairs expect for very low-resource language pairs.
IMO, it is difficult for MT tasks as well.
> 2. Can I use one pre-trained ST model to fine-tune another ST model on a new language pair? Instead of starting ST training from scratch, I wanted to know if it's possible to fine-tune an existing ST model. I'm aware that I can use pretrained ASR and MT models for ST training, but I was wondering if I could directly fine-tune an existing model.


It depends on the demand.
For example, Libri-trans has 100h audio data.
If the language pair is the same, it might be possible to reduce the required training data.
But it's a good research direction.
> 3. Ideally, how many hours of speech data would I need to get a ST model working reasonably well?Will this amount reduce if I'm finetuning on an existing ST model instead of training from scratch?


You can see something you want in
https://arxiv.org/abs/2004.10234
https://arxiv.org/abs/1909.06317
> 4. Could you please guide me to the paper/architecture for the transformer based ST model used in Spanish-English ST task? I was unable to find this and it would be very helpful to know.

**Comment by balag59** [comment]
> @balag59
> If you want to get better models, you need a triplet (source audio, source text, target text).
> 
> > 1. How would one go about training a ST model on a custom dataset with source audio and target text? What format does the data need to be in and what forms of pre-processing would be necessary?
> 
> I'm not sure ST models can be transferred to other language pairs expect for very low-resource language pairs.
> IMO, it is difficult for MT tasks as well.
> 
> > 1. Can I use one pre-trained ST model to fine-tune another ST model on a new language pair? Instead of starting ST training from scratch, I wanted to know if it's possible to fine-tune an existing ST model. I'm aware that I can use pretrained ASR and MT models for ST training, but I was wondering if I could directly fine-tune an existing model.
> 
> It depends on the demand.
> For example, Libri-trans has 100h audio data.
> If the language pair is the same, it might be possible to reduce the required training data.
> But it's a good research direction.
> 
> > 1. Ideally, how many hours of speech data would I need to get a ST model working reasonably well?Will this amount reduce if I'm finetuning on an existing ST model instead of training from scratch?
> 
> You can see something you want in
> https://arxiv.org/abs/2004.10234
> https://arxiv.org/abs/1909.06317
> 
> > 1. Could you please guide me to the paper/architecture for the transformer based ST model used in Spanish-English ST task? I was unable to find this and it would be very helpful to know.

Thank you for the quick response!
1. I do have the triplet of source speech in wav/mp3/flac, source text and target text. How do I train my own ST model on this? I tried looking into the existing ST recipes but the data preparation in each of them seem very specific to the dataset and not something that is easily generalizable to any dataset. How do I prepare this data and train a ST model on this?

2. This is a Chinese-English ST task. What would be the most effective way to pre-train this model? Could I use an existing ST model like Spanish-English as a checkpoint and continue from there? Or do I use a pre-trained ASR model for Chinese like the one on AISHELL for the ST model?

Thanks in advance!

**Comment by hirofumi0810** [comment]
I put my efforts to keep consistency across datasets. It is not so difficult.
Because case information and punctuation are different in each corpus, you need to carefully treat these parts.
> I do have the triplet of source speech in wav/mp3/flac, source text and target text. How do I train my own ST model on this? I tried looking into the existing ST recipes but the data preparation in each of them seem very specific to the dataset and not something that is easily generalizable to any dataset. How do I prepare this data and train a ST model on this?

I personally recommend to use a Chinse ASR encoder and a Zh->En MT decoder.
> This is a Chinese-English ST task. What would be the most effective way to pre-train this model? Could I use an existing ST model like Spanish-English as a checkpoint and continue from there? Or do I use a pre-trained ASR model for Chinese like the one on AISHELL for the ST model?

**Comment by balag59** [comment]
> I put my efforts to keep consistency across datasets. It is not so difficult.
> Because case information and punctuation are different in each corpus, you need to carefully treat these parts.
> 
> > I do have the triplet of source speech in wav/mp3/flac, source text and target text. How do I train my own ST model on this? I tried looking into the existing ST recipes but the data preparation in each of them seem very specific to the dataset and not something that is easily generalizable to any dataset. How do I prepare this data and train a ST model on this?
> 
> I personally recommend to use a Chinse ASR encoder and a Zh->En MT decoder.
> 
> > This is a Chinese-English ST task. What would be the most effective way to pre-train this model? Could I use an existing ST model like Spanish-English as a checkpoint and continue from there? Or do I use a pre-trained ASR model for Chinese like the one on AISHELL for the ST model?

Thanks for the help! Just one last question:
For the Zh-En ST task, which recipe do you recommend following?

**Comment by hirofumi0810** [comment]
@balag59 Unfortunately, there is no Zh->En ST corpus AFAIK.

**Comment by balag59** [comment]
> @balag59 Unfortunately, there is no Zh->En ST corpus AFAIK.

I understand, what I meant to ask was that which of the existing ST recipes would be the easiest to modify to convert it into a Zh-En task? Especially the data part, assuming I have wav files and txt files with their corresponding transcripts and translations. If none of the recipes are a good fit, what do you suggest is the best way to start this as I'm unable to proceed as I'm new to espnet and kaldi style recipes. Thank you for your time!

**Comment by balag59** [comment]
@hirofumi0810 I was able to modify the LibriTrans recipe for Zh-En task and am able to train it so  thank you for your help! I had a couple of follow up questions on using this:

1. The predictions on dev and test are okay and so is the bleu score. However, the train bleu score remains 0 throughout for some reason. Any idea why this is happening?

2. I downloaded the pretrained aishell ASR model and gave that as in input to the encoder init. Is that the right way to initialize the encoder? Will this create an architecture mismatch because of the difference in recipes? 

Thanks!

**Comment by balag59** [comment]
@hirofumi0810 I just had another minor clarification regarding this. I see an option for multi-task training in the train yaml file. How do I use this to train ST with multi task training? In addition, can I pass the pretrained AISHELL ASR model to initialise the encoder here even though it's a different recipe?
Thanks a ton!

**Comment by hirofumi0810** [comment]
@balag59 If you specify the same encoder architecture in the yaml file, you can use the pre-trained AISHELL-1 model as a seed. I think as long as you follow the default recipe, the encoder dimensions should be the same.
Regarding BLEU, I know that issue. I will fix it.

**Comment by balag59** [comment]
> @balag59 If you specify the same encoder architecture in the yaml file, you can use the pre-trained AISHELL-1 model as a seed. I think as long as you follow the default recipe, the encoder dimensions should be the same.
> Regarding BLEU, I know that issue. I will fix it.

@hirofumi0810 Thanks! The issue is that I'm trying to use the AISHELL ASR model with the libritrans ST recipe( There's no ST recipe for AISHELL) so there is a difference in the encoder dimensions and architecture. Is there a workaround for this?

**Comment by rjzevallos** [comment]
Hello @balag59, What is the format of the data folder you changed? What are the files that each folder should have?
